subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/SyncTaskIntegrationTest.groovy

```diff
@@ -394,36 +393,71 @@ class SyncTaskIntegrationTest extends AbstractIntegrationSpec {
         ins.close()
     }
 
-    @ToBeImplemented
     @Issue("https://github.com/gradle/gradle/issues/9586")
-    @Requires(TestPrecondition.CASE_INSENSITIVE_FS)
-    def "change in case of input file still syncs properly"() {
+    def "change in case of input files and folders will still syncs properly"() {
         given:
-        buildFile << """
-            task writeFile {
-                def outputFile = new File(buildDir, project.hasProperty("capitalize") ? "OUTPUT" : "output")
-                outputs.file outputFile
-                doLast {
-                    outputFile.text = "hello"
-                }
-            }
-            task sync(type: Sync) {
-                from writeFile
-                into new File(buildDir, "sync")
+        buildFile << '''
+            task syncIt(type: Sync) {
+                from project.hasProperty("capitalizeFile") ? "FILE.TXT" : "file.txt"
+                from project.hasProperty("capitalizeDir") ? "DIR" : "dir"
+                into buildDir
             }
-        """
+        '''
+
+        def lowercaseFile = file('file.txt')
+        lowercaseFile.text = 'file'
+
+        file('dir').create {
+            file('file2.txt') << 'file2'
+            file('file3.txt') << 'file3'
+            file('file4.txt') << 'file4'
+        }
+
+        and:
+        run 'syncIt'
+        file('build').assertHasDescendants(
+            'file.txt',
+            'file2.txt',
+            'file3.txt',
+            'file4.txt'
+        )
+
+        and:
+        lowercaseFile.delete() // renameTo wouldn't change the case
+        def uppercaseFile = file('FILE.TXT')
+        uppercaseFile.text = 'FILE'
+        assert uppercaseFile.canonicalFile.name == 'FILE.TXT'
+        and:
+        file('dir/file3.txt').delete()
+
+        file('dir/file4.txt').delete()
+        file('dir/File4.txt') << 'File4'
+
         when:
-        succeeds("sync")
+        succeeds('syncIt', '-PcapitalizeFile')
+        def uppercaseOutputFile = file('build/FILE.TXT')
         then:
-        file("build/output").assertExists()
-        file("build/sync/output").assertExists()
+        executedAndNotSkipped ':syncIt'
+        uppercaseOutputFile.parentFile.listFiles() != [].toArray()
+        uppercaseOutputFile.assertExists()
+        uppercaseOutputFile.text == 'FILE'
+        // TODO uppercaseOutputFile.canonicalFile.name == 'FILE.TXT'
```

Comment:
@lptr, should we add some custom logic that if the filename differs, do an actual delete and recreate?

Reply:
Done